### PR TITLE
WIP:Feature/opti error handling

### PR DESCRIPTION
**Pull request by karelrooted**
opti error handling

### CODE REVIEW

File: src/proto/nc_memcache.c [hunk 1]
```diff
@@ -1200,6 +1207,26 @@ memcache_parse_rsp(struct msg *r)
...
+memcache_error_reply(struct msg *r)
+{
+    struct conn *c_conn;
+    struct msg *response = r->peer;
+
+    ASSERT(response != NULL && response->owner != NULL);
+
+    c_conn = response->owner;
+
+    switch (r->result) {
```
Comment (git-hulk) [inline src/proto/nc_memcache.c]: Would Memcached protocol never encounter wrong argument number error?

File: src/nc_request.c [hunk 2]
```diff
@@ -669,6 +669,27 @@ req_recv_done(struct context *ctx, struct conn *conn, struct msg *msg,
         return;
     }
 
+    if (msg->result == MSG_PARSE_ERROR || msg->result == MSG_PARSE_ERROR_WRONG_ARGU_NUM) {
```
Comment (git-hulk) [inline src/nc_request.c]: merge this with noforward?

Comment (karelrooted) [reply src/nc_request.c]: had merged with noforward, also the error_reply func has been merged into reply func,please check the new commit

**Review by git-hulk — APPROVED** [review]

File: src/nc_message.h [hunk 3]
```diff
@@ -29,7 +29,9 @@ typedef bool (*msg_failure_t)(struct msg *r);
 
 typedef enum msg_parse_result {
     MSG_PARSE_OK,                         /* parsing ok */
-    MSG_PARSE_ERROR,                      /* parsing error */
+    MSG_PARSE_ERROR,                      /* parsing error: usually protocol error */
+    MSG_PARSE_ERROR_WRONG_ARGU_NUM,       /* parsing error: wrong number of arguments */
```
Comment (ruoshan) [inline src/nc_message.h]: ARGU => ARG or ARGUMENT

File: src/nc_message.h [hunk 4]
```diff
@@ -228,7 +230,7 @@ struct msg {
     msg_parse_result_t   result;          /* message parsing result */
 
     msg_fragment_t       fragment;        /* message fragment */
-    msg_reply_t          reply;           /* generate message reply (example: ping) */
+    msg_reply_t          reply;           /* generate message (error) reply (example: ping,invalid cmd,wrong argu num)*/
```
Comment (ruoshan) [inline src/nc_message.h]: argu 不常见的

File: src/proto/nc_redis.c [hunk 5]
```diff
@@ -1789,14 +1790,37 @@ redis_parse_req(struct msg *r)
...
+unknown_command_error:
+    r->result = MSG_PARSE_ERROR_UNKNOWN_COMMAND;
+    r->noforward = 1;
+    r->state = state;
+    errno = EINVAL;
+
+    log_hexdump(LOG_INFO, b->pos, mbuf_length(b), "parsed bad req(unknown command) %"PRIu64" "
+                                                                          "res %d type %d state %d", r->id, r->result, r->type,
+                r->state);
+
```
Comment (ruoshan) [inline src/proto/nc_redis.c]: missing `return` here

File: src/nc_response.c [hunk 6]
```diff
@@ -196,12 +196,12 @@ rsp_filter(struct context *ctx, struct conn *conn, struct msg *msg)
     if (msg->failure(msg)) {
         log_debug(LOG_INFO, "server failure rsp %"PRIu64" len %"PRIu32" "
                   "type %d on s %d", msg->id, msg->mlen, msg->type, conn->sd);
-        rsp_put(msg);
 
         conn->err = EINVAL;
         conn->done = 1;
 
-        return true;
+        //continue forward response msg to client before close server
```
Comment (ruoshan) [inline src/nc_response.c]: 注释一般 // 后面加个空格

File: src/proto/nc_redis.c [hunk 7]
```diff
@@ -2791,14 +2816,21 @@ redis_fragment(struct msg *r, uint32_t ncontinuum, struct msg_tqh *frag_msgq)
...
-{
+redis_reply(struct msg *r) {
     struct conn *c_conn;
     struct msg *response = r->peer;
 
     ASSERT(response != NULL && response->owner != NULL);
 
     c_conn = response->owner;
+
+    switch (r->result) {
```
Comment (ruoshan) [inline src/proto/nc_redis.c:2826]: 这里是不是漏了 MSG_PARSE_ERROR 这个 case, 前面把 PARSE_ERROR 也标记为 noforward 了

File: src/nc_message.h (hunk 3 above)
Comment (karelrooted) [reply src/nc_message.h]: fixed

File: src/nc_message.h (hunk 4 above)
Comment (karelrooted) [reply src/nc_message.h]: fixed

File: src/nc_response.c (hunk 6 above)
Comment (karelrooted) [reply src/nc_response.c]: fixed

File: src/proto/nc_redis.c (hunk 7 above)
Comment (karelrooted) [reply src/proto/nc_redis.c:2826]: fixed, wrongly set noforward=1 (setted in a old commit, forgot to remove ), redis MSG_PARSE_ERROR is protocol error, will close the connection in the msg_parse stage, and will not goto req_recv_done func, so we don't need to set noforward=1 in redis_request_parse stage

**Pull request closed without merging**